fcrepo-kernel-impl/src/test/java/org/fcrepo/kernel/impl/services/CreateResourceServiceImplTest.java

```diff
@@ -0,0 +1,264 @@
+
+    @Mock
+    private ResourceHeaders resourceHeaders;
+
+    @Mock
+    private RdfSourceOperationBuilder builder;
+
+    @Mock
+    private RdfSourceOperation operation;
+
+    @Mock
+    private ExternalContent extContent;
+
+    @InjectMocks
+    private CreateResourceServiceImpl createResourceService;
+
+    private final static String txId = "tx1234";
+
+    private final static Collection<String> nonRdfSourceTypes;
+
+    private final static Collection<String> basicContainerTypes;
+
+    private final Model model = ModelFactory.createDefaultModel();
+
+    private final Collection<String> digests = Stream.of("urn:sha1:12345abced")
+            .collect(Collectors.toCollection(HashSet::new));
+
+    static {
+        nonRdfSourceTypes = Collections.singleton(NON_RDF_SOURCE.toString());
+        basicContainerTypes = Collections.singleton(BASIC_CONTAINER.toString());
+    }
+
+    @Before
+    public void setUp() {
+        rdfSourceOperationFactory = new RdfSourceOperationFactoryImpl();
+        setField(createResourceService, "rdfSourceOperationFactory", rdfSourceOperationFactory);
+        nonRdfSourceOperationFactory = new NonRdfSourceOperationFactoryImpl();
+        setField(createResourceService, "nonRdfSourceOperationFactory", nonRdfSourceOperationFactory);
+        when(psManager.getSession(ArgumentMatchers.any())).thenReturn(psSession);
+        when(minter.get()).thenReturn(UUID.randomUUID().toString());
+        when(extContent.getURL()).thenReturn("http://example.org/rest/object");
+        when(extContent.getHandling()).thenReturn(ExternalContent.PROXY);
+        when(extContent.getContentType()).thenReturn("text/plain");
+    }
+
+    @Test(expected = ItemNotFoundException.class)
+    public void testNoParentRdf() throws Exception {
+        final String fedoraId = UUID.randomUUID().toString();
+        when(psSession.getHeaders(fedoraId, null)).thenThrow(PersistentItemNotFoundException.class);
+        createResourceService.perform(txId, fedoraId, null, null, null, model);
+    }
+
+    @Test(expected = CannotCreateResourceException.class)
+    public void testParentIsBinaryRdf() throws Exception {
+        final String fedoraId = UUID.randomUUID().toString();
+        when(resourceHeaders.getTypes()).thenReturn(nonRdfSourceTypes);
+        when(psSession.getHeaders(fedoraId, null)).thenReturn(resourceHeaders);
+        createResourceService.perform(txId, fedoraId, null, null, null, model);
+
+    }
+
+    @Test
+    public void testSlugIsNullRdf() throws Exception {
+        final String fedoraId = UUID.randomUUID().toString();
+        when(psSession.getHeaders(fedoraId, null)).thenReturn(resourceHeaders);
+        when(resourceHeaders.getTypes()).thenReturn(basicContainerTypes);
+        createResourceService.perform(txId, fedoraId, null, null, null, model);
+        verify(psSession).persist(ArgumentMatchers.any(ResourceOperation.class));
+    }
+
+    @Test
+    public void testWithSlugExistsRdf() throws Exception {
```

Comment:
Might make sense to verify that the resource operation persisted doesn't have an internal id with the slug in it. You can declare a ArgumentCaptor for the test class and use it to get the ResourceOperation persisted back. 
```
    @Captor
    private ArgumentCaptor<ResourceOperation> operationCaptor;
```